citest/base/bindings.py

```diff
@@ -0,0 +1,376 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Support for bindings and configuration."""
+
+
+import argparse
+import copy
+import inspect
+import os
+
+import ConfigParser
+
+def _normalize_key(key):
+  """Handle historic interpretation of binding keys being upper-cased.
+
+  Converts the queries to lower case. This does not enforce that keys added
+  through configuration files are lower cased, in which case they will not
+  be findable. Eventuallyt this should be the identity.
+  """
+  return key.lower()
+
+
+class ConfigurationBindings(object):
+  """Represents variable bindings for a configuration.
+
+  ConfigurationBindings is essentially a dictionary of name/value pairs.
+  However they are composed from a hierarchical search space and make use
+  of python "ConfigParser" to specify them in a static file.
+
+  The search space is (in order of precedence):
+    Overrides - These are explicit values injected at runtime.
+    ConfigParser - These are values read from configuration files[*].
+    Defaults - These are static default values if not found in a configuration
+       file.
+
+
+  The ConfigParser has an additional hierarchy of sections. If the instance
+  section attribute is set, then only that section in the ConfigParser will be
+  considerd. The Defaults are implemented as part of the ConfigParser
+
+  Bindings are typically collected and assembled by a
+  ConfigurationBindingsBuilder.
+  """
+
+  @property
+  def config_parser(self):
+    """Returns the underlying ConfigParser."""
+    return self.__config_parser
+
+  @property
+  def overrides(self):
+    """Returns the underlying overrides dictionary."""
+    return self.__overrides
+
+  def __init__(self, config_parser, overrides, section=None):
+    """Construct a bindings instance.
+
+    Args:
+      config_parser: [ConfigParser] Retrieves values from configuration files.
+      overrides: [dict] Values that override anything found in config_parser.
```

Comment:
Updated the Overrides section in the class docstring. Flags are overrides, but there could be other mechanisms for this as well, like an explicit assignment.